### PR TITLE
perf(subscriptions): Avoid slow delegate task builder where possible

### PR DESCRIPTION
**Pull request by lynnagara**
The delegate task builder is much slower than using the immediate
or jittered one since it has to check which mode to apply for
every subscription within the tick individually. This change uses
the immediate or jittered task builder directly for the whole tick
if we are not in a transition mode. This is checked at the start
of each tick.

### CODE REVIEW

**Comment by lynnagara** [comment]
> Can you provide some background on what these builders do?

There are 2 main builders - immediate and jittered. Immediate means that the subscription is scheduled "immediately", i.e. if its resolution is 1 minute queries are run at the start of the minute, if 1 hour at the start of each hour etc. The problem we ran into is that it causes a huge load on ClickHouse at the start of each minute and hour, which is why the jittered builder was introduced. It delays the execution of each subscription by some stable jitter amount based on the subscription ID. The delegate builder supports transitioning between the modes.

File: snuba/subscriptions/scheduler.py [hunk 1]
```diff
@@ -323,6 +347,7 @@ def __get_subscriptions(self) -> List[Subscription]:
         return self.__subscriptions
 
     def find(self, tick: Tick) -> Iterator[ScheduledSubscriptionTask]:
+        self.__reset_builder()
```
Comment (fpacifici) [inline snuba/subscriptions/scheduler.py:350]: This is executed once per message on the commit log, right ?

**Review by fpacifici — APPROVED** [review]

File: snuba/subscriptions/scheduler.py (hunk 1 above)
Comment (lynnagara) [reply snuba/subscriptions/scheduler.py:350]: Yes, that's how the ticks get created. One per per message.

**Comment by onewland** [comment]
> > Can you provide some background on what these builders do?
> 
> There are 2 main builders - immediate and jittered. Immediate means that the subscription is scheduled "immediately", i.e. if its resolution is 1 minute queries are run at the start of the minute, if 1 hour at the start of each hour etc. The problem we ran into is that it causes a huge load on ClickHouse at the start of each minute and hour, which is why the jittered builder was introduced. It delays the execution of each subscription by some stable jitter amount based on the subscription ID. The delegate builder supports transitioning between the modes.

and what is a builder? what does it do?

**Pull request merged**